### PR TITLE
compare atomic and mutex performance in case of contention.

### PR DESCRIPTION
**Pull request by lyuxuan**

### CODE REVIEW

**Comment by lyuxuan** [comment]
BenchmarkStoreContentionWithAtomic-12    	2000000000	         1.38 ns/op
BenchmarkStoreContentionWithMutex-12     	1000000000	         2.40 ns/op

File: benchmark/primitives/primitives_test.go [hunk 1]
```diff
@@ -277,56 +276,30 @@ func BenchmarkAtomicTimePointerStore(b *testing.B) {
...
-					wg.Done()
-				}()
-			}
-			_ = c
-			wg.Wait()
-		})
-	}
+	b.SetParallelism(10000)
+	b.RunParallel(func(pb *testing.PB) {
+		var c unsafe.Pointer
```
Comment (coocood) [inline benchmark/primitives/primitives_test.go]: The variable is defined in each goroutine, so there is no contention.

File: benchmark/primitives/primitives_test.go [hunk 2]
```diff
@@ -277,56 +276,30 @@ func BenchmarkAtomicTimePointerStore(b *testing.B) {
...
+			atomic.StorePointer(&c, unsafe.Pointer(&t))
+		}
+	})
+}
+
+func BenchmarkStoreContentionWithMutex(b *testing.B) {
+	t := 123
+	b.SetParallelism(10000)
+	b.RunParallel(func(pb *testing.PB) {
+		var c int
```
Comment (coocood) [inline benchmark/primitives/primitives_test.go]: ditto

File: benchmark/primitives/primitives_test.go (hunk 1 above)
Comment (lyuxuan) [reply benchmark/primitives/primitives_test.go]: Thanks for pointing out.

**Comment by dfawley** [comment]
Looks like there is some kind of lint error still.  Also, can you update the benchmark numbers please?

**Comment by lyuxuan** [comment]
```
BenchmarkStoreContentionWithAtomic-12    	100000000	        15.8 ns/op
BenchmarkStoreContentionWithMutex-12     	20000000	       116 ns/op
```
```
BenchmarkStructStoreContentionWithCAS/CAS/100000/100000000-12         	       1	1421414467 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/100000/10000-12             	 2000000	       608 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/100000/0-12                 	10000000	       174 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/10000/100000000-12          	     300	   4822022 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/10000/10000-12              	 3000000	       490 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/10000/0-12                  	10000000	       160 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/1000/100000000-12           	     300	   4771404 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/1000/10000-12               	 3000000	       484 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/1000/0-12                   	10000000	       162 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/10/100000000-12             	     300	   4888170 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/10/10000-12                 	 3000000	       512 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/10/0-12                     	10000000	       168 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/1/100000000-12              	     300	   4891607 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/1/10000-12                  	 3000000	       482 ns/op
BenchmarkStructStoreContentionWithCAS/CAS/1/0-12                      	10000000	       163 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/100000/100000000-12     	     200	   5778961 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/100000/10000-12         	 2000000	      2876 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/100000/0-12             	 5000000	       205 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/10000/100000000-12      	     200	   5790696 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/10000/10000-12          	 1000000	      3148 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/10000/0-12              	10000000	       265 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/1000/100000000-12       	     300	   4830805 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/1000/10000-12           	 1000000	      1896 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/1000/0-12               	10000000	       452 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/10/100000000-12         	     300	   4861557 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/10/10000-12             	 3000000	       497 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/10/0-12                 	20000000	       105 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/1/100000000-12          	     300	   4927593 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/1/10000-12              	 3000000	       488 ns/op
BenchmarkStructStoreContentionWithMutex/Mutex/1/0-12                  	10000000	       136 ns/op
```

**Comment by dfawley** [comment]
> XXXX(set by SetParallelism)/XXXX (extra work besides critical section)

Could you reverse the order of these so when the tests are printed, they're easier to compare?

Also, please either remove the extra "/Mutex" and "/CAS" from the name, or merge the two benchmarks into a single top-level benchmark function.

File: benchmark/primitives/primitives_test.go [hunk 3]
```diff
@@ -277,55 +277,88 @@ func BenchmarkAtomicTimePointerStore(b *testing.B) {
...
-		b.Run(fmt.Sprintf("AtomicStorePointer/%v", n), func(b *testing.B) {
-			var wg sync.WaitGroup
-			var up unsafe.Pointer
-			for i := 0; i < n; i++ {
-				wg.Add(1)
-				go func() {
-					for j := 0; j < b.N; j++ {
-						atomic.StorePointer(&up, unsafe.Pointer(&t))
+					for pb.Next() {
+						for y := 0; y < j; y++ {
```
Comment (dfawley) [inline benchmark/primitives/primitives_test.go]: I believe you can do: `for y := 0; y < j; y++ {}` and the compiler won't remove it.

File: benchmark/primitives/primitives_test.go [hunk 4]
```diff
@@ -277,55 +277,88 @@ func BenchmarkAtomicTimePointerStore(b *testing.B) {
...
+type dummyStruct struct {
+	a int64
+	b time.Time
+}
+
+func BenchmarkStructStoreContentionWithCAS(b *testing.B) {
+	d := dummyStruct{}
+	dp := unsafe.Pointer(&d)
+	t := time.Now()
+	for _, i := range []int{100000, 10000, 1000, 10, 1} {
```
Comment (dfawley) [inline benchmark/primitives/primitives_test.go]: Can we remove one or two of these?  Maybe 10k (or 100k) and 10?

**Comment by lyuxuan** [comment]
```
BenchmarkStructStoreContention/CAS/100000000/100000-12         	       1	1533096236 ns/op
BenchmarkStructStoreContention/CAS/10000/100000-12             	 2000000	       582 ns/op
BenchmarkStructStoreContention/CAS/0/100000-12                 	10000000	       173 ns/op
BenchmarkStructStoreContention/CAS/100000000/10-12             	     300	   4908433 ns/op
BenchmarkStructStoreContention/CAS/10000/10-12                 	 3000000	       492 ns/op
BenchmarkStructStoreContention/CAS/0/10-12                     	10000000	       168 ns/op
BenchmarkStructStoreContention/Mutex/100000000/100000-12       	     200	   5719244 ns/op
BenchmarkStructStoreContention/Mutex/10000/100000-12           	 2000000	      3023 ns/op
BenchmarkStructStoreContention/Mutex/0/100000-12               	 2000000	       528 ns/op
BenchmarkStructStoreContention/Mutex/100000000/10-12           	     300	   5231189 ns/op
BenchmarkStructStoreContention/Mutex/10000/10-12               	 3000000	       474 ns/op
BenchmarkStructStoreContention/Mutex/0/10-12                   	10000000	       100 ns/op
```

**Comment by dfawley** [comment]
Looks good, but could you reverse the order of the parameters so when the tests are printed, they're easier to compare?  I.e. put the iteration count before the parallelism.  The output would look like this instead (but with the /x/y/ numbers reordered to /y/x/, which I did not want to do):

```
BenchmarkStructStoreContention/CAS/100000000/100000-12         	    3000	   4731686 ns/op
BenchmarkStructStoreContention/CAS/100000000/10-12             	    3000	   4677181 ns/op
BenchmarkStructStoreContention/CAS/10000/100000-12             	30000000	       483 ns/op
BenchmarkStructStoreContention/CAS/10000/10-12                 	30000000	       475 ns/op
BenchmarkStructStoreContention/CAS/0/100000-12                 	100000000	       155 ns/op
BenchmarkStructStoreContention/CAS/0/10-12                     	100000000	       154 ns/op
BenchmarkStructStoreContention/Mutex/100000000/100000-12       	    3000	   4745838 ns/op
BenchmarkStructStoreContention/Mutex/100000000/10-12           	    3000	   4687433 ns/op
BenchmarkStructStoreContention/Mutex/10000/10-12               	30000000	       479 ns/op
BenchmarkStructStoreContention/Mutex/10000/100000-12           	 5000000	      3442 ns/op
BenchmarkStructStoreContention/Mutex/0/100000-12               	30000000	       365 ns/op
BenchmarkStructStoreContention/Mutex/0/10-12                   	200000000	       100 ns/op
```

(Note how it's easier to see how the workload scales with less/more parallelism, because the relevant lines are adjacent.)

**Comment by dfawley** [comment]
Also.. I assume this is an anomaly?

> `BenchmarkStructStoreContention/Mutex/10000/10-12               	30000000	       479 ns/op`
> `BenchmarkStructStoreContention/Mutex/10000/100000-12           	 5000000	      3442 ns/op`

The Mutex doesn't scale that poorly for the other workloads (with higher or lower contention).

**Comment by lyuxuan** [comment]
Well, it's not anomaly, quite repeatable. I also remembered it's of magnitude 1000 when we did experiments before.

```
BenchmarkStructStoreContention/CAS/100000000/100000-12         	    3000	   4808601 ns/op
BenchmarkStructStoreContention/CAS/100000000/10-12             	    3000	   4694737 ns/op
BenchmarkStructStoreContention/CAS/10000/100000-12             	30000000	       482 ns/op
BenchmarkStructStoreContention/CAS/10000/10-12                 	30000000	       475 ns/op
BenchmarkStructStoreContention/CAS/0/100000-12                 	100000000	       155 ns/op
BenchmarkStructStoreContention/CAS/0/10-12                     	100000000	       153 ns/op
BenchmarkStructStoreContention/Mutex/100000000/100000-12       	    3000	   4757765 ns/op
BenchmarkStructStoreContention/Mutex/100000000/10-12           	    3000	   4680524 ns/op
BenchmarkStructStoreContention/Mutex/10000/100000-12           	 3000000	      3353 ns/op
BenchmarkStructStoreContention/Mutex/10000/10-12               	20000000	       665 ns/op
BenchmarkStructStoreContention/Mutex/0/100000-12               	50000000	       430 ns/op
BenchmarkStructStoreContention/Mutex/0/10-12                   	100000000	       111 ns/op
```

**Comment by dfawley** [comment]
> Well, it's not anomaly, quite repeatable. I also remembered it's of magnitude 1000 when we did experiments before.

Yes...it looks like we have a 4x slowdown between the 10 & 100k parallelism for the max-contention case as well, so this is on par with that.

**Pull request merged**